### ISSUE TITLE
Game hangs with message "hang on a bit..." constantly

### ISSUE THREAD

**Issue by schealex** [CleverRaven/Cataclysm-DDA#63202]
### Describe the bug

In my last two created words i get this weird behavior that during long activities (reading, sleeping, hacksawing etc.) almost always there is this message on top of the screen "hang on a bit..." which takes like ~5 real world minutes to go away. Game is literally unplayable like this. I've tried updating to the latest experimental 2023-01-19-0645 and also reinstalling the game completly but the problem is still there. I'm attaching my save file for the current world (still small as just started) and my debug.log (which doesn't seem to contain anything regarding this behavior).

[Sistersville.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/10464427/Sistersville.zip)
[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/10464435/debug.log)

### Steps to reproduce

1. Load the save
2. Try long activities (hacksaw metal bars, sleeping etc)

### Expected behavior

Let me play without wasting 5 minutes of my time every couple minutes.

### Versions and configuration

- Win 10
- Tiles
- Version: 2023-01-19-0645
- Lang: German
- Mods: "dda", "no_npc_food", "no_fungal_growth", "package_bionic_professions", "blazeindustries", "speedydex", "StatsThroughSkills", "personal_portal_storms", "stats_through_kills"

**Comment by Brambor** [comment]
It did happen to me too (once) recently.

**Comment by alp7292** [comment]
its so annoying when i try to sleep game says hang on a bit and takes 5 minute whats more is that if i altf4 and load the save before sleeping and sleep again it doesnt happen and saves 4 minute

**Comment by andrei8l** [comment]
Probably caused by NPCs moving on the overmap due to #62985

https://github.com/CleverRaven/Cataclysm-DDA/blob/02df847acfe3d5ab8d1cd5526e7708fb7b7e8bfa/src/do_turn.cpp#L572-L573

https://github.com/CleverRaven/Cataclysm-DDA/blob/02df847acfe3d5ab8d1cd5526e7708fb7b7e8bfa/src/overmapbuffer.cpp#L982-L983

though I haven't debugged it yet.

**Comment by schealex** [comment]
> Probably caused by NPCs moving on the overmap due to #62985
> 
> https://github.com/CleverRaven/Cataclysm-DDA/blob/02df847acfe3d5ab8d1cd5526e7708fb7b7e8bfa/src/do_turn.cpp#L572-L573
> 
> https://github.com/CleverRaven/Cataclysm-DDA/blob/02df847acfe3d5ab8d1cd5526e7708fb7b7e8bfa/src/overmapbuffer.cpp#L982-L983
> 
> though I haven't debugged it yet.

i'll downgrade to before experimental-2023-01-09-0348 to test!
but that sure sounds like it is related to this issue!

**Comment by schealex** [comment]
ok so i've tested a bit and it's definitly not caused by the changes of #62985 because it's now still doing it but without the "hang on a bit..." message. There must be a change before that which triggered this behavior

**Comment by irwiss** [comment]
Your world has NPC_SPAWN_RATE set to 0.01, this practically guarantees that every (in game) hour an npc will attempt to spawn. 

If the (randomly) chosen goal wants to do overmap pathfinding you'll see either the spinner (after #62985) or just a frozen window on a commit before that PR

Set your npc spawn rate to default (4.00) in world options and it'll be a significantly rarer event, but you still occasionally get to see the spinner, when random npc needs to pathfind to their goal

Here's what happens when you just wait 24 hours on "fresh" world with no overmap explored with different npc spawn rates
![image](https://user-images.githubusercontent.com/6560075/213811701-c440482c-270e-4981-ac7a-1302fb1c8e1e.png)

**Comment by schealex** [comment]
@irwiss okay I get that but why is that a problem all the sudden? I've played multiple worlds with the exact same settings before and never encountered this issue. Also shouldnt the game prevent you from setting it this low if there is a problem with it being that now?

**Comment by irwiss** [comment]
If you have the time (and the will) you could test 0.01 spawn rate in older versions that you used and trace which version introduced it... But it's possible it'll go back to either when the random spawn was introduced, when npc missions started using pathfinding or similar...

Don't know why "all of a sudden" it's a problem as the source of the 'sudden' isn't traced, FWIW the game does hide these settings under "advanced", the basic/modern worldgen slider only allows to go down to 1.00

The spinner is stopgap solution so the game isn't completely frozen during pathfinding, the "correct" solution would be to kick the pathfinding off the main thread, or otherwise optimize it to be at least an order of magnitude faster, not sure if either solution will get merged in the 0.G window even if proposed

**Comment by schealex** [comment]
ok so i raised the npc spawn to 2.00 now but that just lengthens the time between the "hang on a bit..." stuff. I still cannot believe that it's meant to take like ~5 real world minutes if that event occurs. There has to be something wrong with it?!

**Comment by anoobindisguise** [comment]
I'm also having this problem and only as of very very recently.

**Comment by Nebnis** [comment]
I'm also having this problem on a world with normal npc spawn and very sparse cities. My game is locking up quite often for like 10 minutes instead during activities

**Comment by mqrause** [comment]
Pretty much a duplicate of #54683, except now there's the message instead of just hanging.

**Comment by schealex** [comment]
@mqrause yep seems that way but what kind of baffels me is that i've played multiple worlds in 2022 without having this issue even once. So there must be some factor triggering this behavior :/ I'd also say thats not a performance issue but rather a bug.

**Comment by andrei8l** [comment]
Does the `Hang on a bit...` spinner move or is it a frozen image?

**Comment by schealex** [comment]
> Does the `Hang on a bit...` spinner move or is it a frozen image?

It moves generally but occasionally hangs for a couple seconds.
If this is about pathfinding (i still don't get why pathfinding is relevant to npc spawning) i've got a pretty beefy machine -> can't image how long those hang's are for poeple on slower hardware.. :/

**Comment by andrei8l** [comment]
I can reproduce this from the msvc build but not from the linux or mingw builds. Simply add a large construction zone for shallow pits from the save file in the OP and the game will eventually get stuck on the `Hang on a bit...` animation. I'm afraid I can't debug further since it doesn't happen natively.

**Comment by Nebnis** [comment]
> Does the `Hang on a bit...` spinner move or is it a frozen image?

It spins for me, and 5 min is certainly not a gross exaggeration. I got a pretty beefy machine too and while it clears fast in less than 5 min sometimes, i had happen 2 or 3 times where it got stuck quite a while(from like 5~6 times total

I tried debuggging to see if there is a lab nearby but i don't see anything in my fov. I do have wandering hordes on

**Comment by ChromePoptart** [comment]
This just happened to me for the first time.  Went to sleep around 8:30pm with the game just showing "Wait until you wake up."  At 10:30pm, the "Hang on a bit..." spinner popped up and stopped the progression of in-game time for a few real-life minutes.

I don't know if this is at all an influencing factor, but it happened on Autumn day 91, meaning the game transitioned seasons to Winter as 12:00am (1.5 in-game hours after the hanging occurred).  I will also add that the place I was sleeping definitely didn't have any NPCs within a fairly large radius (at least 10 overmap tiles, probably more).

- OS: Windows
    - OS Version: 10.0.19044.2486 (21H2)
- Game Version: 7921045 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions],
    Blaze Industries [blazeindustries],
    Bionic Slots [cbm_slots],
    Portal Storms Ignore NPCs [personal_portal_storms]
]

**Comment by mqrause** [comment]
Profile of [find_overmap_path](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/simple_pathfinding.cpp#L277) while I left it running for a bit on `hang on a bit...`.

![grafik](https://user-images.githubusercontent.com/38702195/219509377-286dba56-0448-4d7a-99db-123ce77307e6.png)

Additional info I have is that it always runs to `max_search_count`, so until `known_nodes` contains 100,000 nodes. The issue seems to be accessing `std::unordered_map<node_address, navigation_node, node_address_hasher> known_nodes`, but that's outside my expertise. Hope it helps someone who knows more about this stuff.
